# <h2><b> Đồ án thực hành </b></h2>
<h1 style='color:red;text-align:center;font-size:35px'><b>Chủ đề: Dự đoán giá rượu vang</b></h1>

# <h3>Các thành viên trong nhóm</h3>
<li> Phan Xuân Hoài - 20120481 </li>
<li> Lê Kinh Luân - 1612355 </li>
<li> Nguyễn Minh Nghĩa - 20120534</li>
<li> Phạm Quốc Hưng - 19120522</li>

# <h1 style='color:blue;text-align:center;font-size:30px'><b>Phần I - B.Khám phá dữ liệu + Tiền xử lý</b></h1>

## Import những thư viện cần thiết

In [1]:
%matplotlib inline
import matplotlib.pyplot as plt
import numpy as np
import pandas as pd 
import seaborn as sns

# YOUR CODE HERE (OPTION)
# Nếu cần các thư viện khác thì bạn có thể import ở đây
import os
import re
import statistics
import seaborn as sns
sns.set(rc={'axes.facecolor':'lightblue', 'figure.facecolor':'lightgray'})

In [ ]:
import sys
sys.executable

In [ ]:
cd champagne/champagne

### Đọc file 'champagnes_info.csv' vào dataframe df

In [4]:
df = pd.read_csv('./dataset/champagnes_info.csv')
df.head(20)

Ma_sp                                           Ten_sp        Gia  \
0     WVN-9400           Rượu Champagne Armand De Brignac Blanc  9.400.000   
1    VNHM-1350                   Rượu Champagne Arnaud Tabourin  1.350.000   
2   VNHM1-1750          Rượu Champagne Autreau Brut Premier Cru  1.750.000   
3   VNHM1-1850    Rượu Champagne Autreau Reserve Brut Grand Cru  1.850.000   
4   VNTK1-2760             Rượu Champagne Ayala Blanc De Blancs  2.760.000   
5    VNTK-2294                 Rượu Champagne Ayala Brut Majeur  2.294.000   
6    VNTK-2734                 Rượu Champagne Ayala Brut Nature  2.734.000   
7    VNPE-1250             Rượu Champagne Bernard Lonclas Blanc  1.250.000   
8   VNWH1-4200  Rượu Champagne Billecart Salmon Blanc De Blancs  2.709.000   
9   VNWH1-2750     Rượu Champagne Billecart Salmon Brut Réserve  1.895.000   
10   VNWH-3900   Rượu Champagne Billecart Salmon Brut Sous Bois  3.900.000   
11    VNDL-552          Rượu Champagne Bisol BelStar Cuvee Rose    495.000   
12  VNTK1-3867     Rượu Champagne Bollinger La Cote Aux Enfants  3.867.000   
13   VNTK-4920         Rượu Champagne Bollinger La Grande Année  4.920.000   
14   VNTK-5960    Rượu Champagne Bollinger La Grande Année Rose  5.960.000   
15   VNTK-3880                    Rượu Champagne Bollinger Rose  3.880.000   
16   VNTK-4134       Rượu Champagne Bollinger Rose Limited 2006  4.134.000   
17   VNTK-3294           Rượu Champagne Bollinger Special Cuvee  3.294.000   
18   VNPV-3958         Rượu Champagne Bottega Stardust Prosecco  3.563.000   
19   VNPV - 11                    Rượu Champagne Bottega Stella          1   

     Xuat_xu     Nong_do   Dung_tich                             Giong_nho  \
0   Pháp\r\n  12.5 %\r\n  750 ML\r\n                            Chardonnay   
1   Pháp\r\n    12 %\r\n  750 ML\r\n                            Chardonnay   
2   Pháp\r\n    12 %\r\n  750 ML\r\n                            Chardonnay   
3   Pháp\r\n    12 %\r\n  750 ML\r\n                 Chardonnay,Pinot Nero   
4   Pháp\r\n    12 %\r\n  750 ML\r\n                            Chardonnay   
5   Pháp\r\n    12 %\r\n  750 ML\r\n   Chardonnay,Pinot Noir,Pinot Meunier   
6   Pháp\r\n    12 %\r\n  750 ML\r\n   Chardonnay,Pinot Noir,Pinot Meunier   
7   Pháp\r\n    12 %\r\n  750 ML\r\n  Pinot Noir,Pinot Meunieur,Chardonnay   
8   Pháp\r\n  12.5 %\r\n  750 ML\r\n                           Chardnonnay   
9   Pháp\r\n  12.5 %\r\n  750 ML\r\n   Chardonnay,Pinot Noir,Pinot Meunier   
10  Pháp\r\n  12.5 %\r\n  750 ML\r\n   Chardonnay,Pinot Noir,Pinot Meunier   
11     Ý\r\n  11.5 %\r\n  750 ML\r\n                     Merlot,Pinot Nero   
12  Pháp\r\n  12.5 %\r\n  750 ML\r\n                            Pinot Noir   
13  Pháp\r\n    12 %\r\n  750 ML\r\n                 Pinot Noir,Chardonnay   
14  Pháp\r\n    14 %\r\n  750 ML\r\n   Pinot Noir,Chardonnay,Pinot Meunier   
15  Pháp\r\n    12 %\r\n  750 ML\r\n                 Chardonnay,Pinot Noir   
16  Pháp\r\n    12 %\r\n  750 ML\r\n                 Chardonnay,Pinot Noir   
17  Pháp\r\n    12 %\r\n  750 ML\r\n                 Chardonnay,Pinot Noir   
18     Ý\r\n    11 %\r\n  750 ML\r\n                                 Glera   
19     Ý\r\n    11 %\r\n  750 ML\r\n                                  \r\n   

               Mau_sac         Loai_ruou           Thuong_hieu Vung_lam_vang  
0            Màu trắng    Rượu Champagne     Armand De Brignac           NaN  
1                         Rượu Champagne       Arnaud Tabourin           NaN  
2                  NaN  Rượu Vang Trắng                   NaN     Champagne  
3                         Rượu Champagne               Autreau           NaN  
4          Vàng nhạt   Rượu Champagne               : Ayala           NaN  
5            Vàng rơm   Rượu Champagne               : Ayala           NaN  
6            Vàng rơm   Rượu Champagne               : Ayala           NaN  
7             Vàng rơm    Rượu Champagne                   NaN           NaN  
8                  N

# Khám phá dữ liệu

### Xem số dòng cột của dữ liệu 

In [5]:
num_rows = df.shape[0]
num_cols = df.shape[1]
shape = (num_rows, num_cols)
shape

(5598, 11)

### Kiểu dữ liệu của các cột 

In [6]:
df.dtypes

Ma_sp            object
Ten_sp           object
Gia              object
Xuat_xu          object
Nong_do          object
Dung_tich        object
Giong_nho        object
Mau_sac          object
Loai_ruou        object
Thuong_hieu      object
Vung_lam_vang    object
dtype: object

### Ý nghĩa của các cột

Đọc file "champagnes_schema.csv" vào DataFrame `col_meaning_df`

In [7]:
temp = pd.read_csv("./dataset/champagnes_schema.csv")
temp
data = {
    "name": temp["name"],
    "mean": temp["mean"]
}
pd.set_option('max_colwidth', 400)


col_meaning_df = pd.DataFrame(data = data)
col_meaning_df = col_meaning_df.set_index(['name'])

In [8]:
col_meaning_df

mean
name                                      
Ma_sp                          Ma san pham
Ten_sp                            Ten ruou
Gia                      Gia tien san pham
Xuat_xu         Xuat xu cua ruou(quoc gia)
Nong_do                   Nong do cua ruou
Dung_tich              Dung tich chai ruou
Giong_nho               Giong nho lam ruou
Mau_sac                   Mau sac cua ruou
Loai_ruou     San pham thuoc loai ruou nao
Thuong_hieu        Thuong hieu lam ra ruou
Vung_lam_nho       Vung san xuat ruou vang

# Tiền xử lý

Loại bỏ các ký tự '\n' (hoặc '\r\n' trong Windows) và các ký tự dư thừa trong các cột 'Xuat_xu', 'Nong_do', 'Dung_tich', 'Giong_nho','Thuong_hieu' để có thể chuyển đổi các cột về dạng số dễ dàng hơn.

In [9]:
def is_nan(x):
    return (x != x)

In [10]:
cate_cols = df.select_dtypes(include = ['object']).columns.tolist()
df[cate_cols] = df[cate_cols].apply(lambda x : x.replace('\\r?\\n', '', regex = True))
df['Nong_do'] = df['Nong_do'].str.replace('%', '')
df['Dung_tich'] = df['Dung_tich'].str.replace('ML', '')
cols = ['Giong_nho', 'Mau_sac', 'Thuong_hieu', 'Vung_lam_vang']
df[cols] = df[cols].apply(lambda x : x.str.replace(':', ''))
df = df.apply(lambda x : x.str.strip())


Chuyển kiểu dữ liệu của các cột 'Dung_tich', 'Nong_do','Gia' về dạng số

In [11]:
df['Nong_do'] = pd.to_numeric(df['Nong_do'], errors='coerce')
df['Dung_tich'] = pd.to_numeric(df['Dung_tich'], downcast='float', errors='coerce')
df['Gia'] = df['Gia'].replace('[.]', '', regex=True).astype(float)

In [12]:
df.dtypes

Ma_sp             object
Ten_sp            object
Gia              float64
Xuat_xu           object
Nong_do          float64
Dung_tich        float32
Giong_nho         object
Mau_sac           object
Loai_ruou         object
Thuong_hieu       object
Vung_lam_vang     object
dtype: object

Lọc đi các chai rượu có giá thấp bất thường 

In [13]:
df.loc[df['Gia'] < 1000, 'Gia'].value_counts()

1.0      408
770.0      1
86.0       1
Name: Gia, dtype: int64

Có tới hơn 400 chai có giá thấp hơn 1000, một mức giá vô lý. Ta sẽ tiến hành lọc đi các chai rượu này.

In [14]:
df = df[~(df['Gia'] < 1000)].reset_index(drop = True)

## Quay lại bước khám phá dữ liệu 

Với các cột có kiểu dữ liệu số, ta cần tính:
- Tỉ lệ % (từ 0 đến 100) các giá trị thiếu 
- Giá trị min
- Giá trị lower quartile (phân vị 25)
- Giá trị median (phân vị 50)
- Giá trị upper quartile (phân vị 75)
- Giá trị max

Sau đó lưu kết quả vào DataFrame `num_col_info_df`, trong đó: 
- Tên của các cột là tên của các cột số trong `df`
- Tên của các dòng là: "missing_ratio", "min", "lower_quartile", "median", "upper_quartile", "max"  

Để dễ nhìn, tất cả các giá trị bạn đều làm tròn với 1 chữ số thập phân bằng phương thức `.round(1)`.

In [15]:
items = ["Gia","Nong_do","Dung_tich"]
percentMissingList = []
for item in items:
   percentMissingList.append(round(df[item].apply(is_nan).tolist().count(1)/shape[0] * 100,1))

percentMissingList

columns = ["Gia","Nong_do","Dung_tich"]
row_labels = ["missing_ratio","min","lower_quartile","median","upper_quartile","max"]
temp = df[["Gia","Nong_do","Dung_tich"]]
info_df = temp.quantile(q=[0.25,0.5, 0.75], axis=0, numeric_only=True, interpolation='midpoint')
max = round(temp.max(axis = 0),1)
min = round(temp.min(axis = 0),1)

data = [percentMissingList,min,info_df.loc[0.25].tolist(),info_df.loc[0.5].tolist(),info_df.loc[0.75].tolist(),max]
num_col_info_df = pd.DataFrame(data = data, index = row_labels, columns = columns)
num_col_info_df

Gia  Nong_do  Dung_tich
missing_ratio          0.0      0.1        0.0
min               105000.0      0.0      187.0
lower_quartile    459000.0     13.0      750.0
median            777000.0     13.5      750.0
upper_quartile   1518000.0     14.0      750.0
max             39000000.0     45.0     6000.0

### Tiền xử lý

Ngoài các giá trị NaN trong cột `Thuong_hieu` còn có các giá trị chuỗi `'', ' ', ':', ...` không hợp lệ. Ta sẽ đặt chúng về giá trị Nan

In [16]:
def fun(x):
    if type(x) == float:
        return True
    return len(x) > 2

In [17]:
mask = df['Thuong_hieu'].apply(fun)
df[~mask].shape

(27, 11)

In [18]:
# mask = df['Thuong_hieu'].apply(fun)
# df.drop(df[~mask].index, inplace=True)
df[~mask] = np.NaN

----

Trong cột `Mau_sac` còn có các giá trị `''`. Ta chuyển giá trị này về Nan

In [19]:
mask = df['Mau_sac'] == ''
df.loc[mask,'Mau_sac'] = np.NaN

----

Xử lý chuỗi của cột "Màu sắc" thành chuỗi viết thường (Vd: ở cột màu săc có "Vàng Rơm", "Vàng rơm"... thì chuyển thành "vàng rơm"), xóa khoảng trắng đầu và cuối, đồng bộ ký tự màu sắc ("màu vàng rơm", "vàng rơm" --> "vàng rơm")

In [20]:
df['Mau_sac'] = df['Mau_sac'].str.lower()
df['Mau_sac'] = df['Mau_sac'].str.strip()

df = df.apply(lambda x: x.replace({'màu':'', ': ':'','vàng':'vàng','đỏ':'đỏ'}, regex=True))
df['Mau_sac'] = df['Mau_sac'].str.strip()


In [21]:
df.head(20)

Ma_sp                                           Ten_sp        Gia  \
0     WVN-9400           Rượu Champagne Armand De Brignac Blanc  9400000.0   
1    VNHM-1350                   Rượu Champagne Arnaud Tabourin  1350000.0   
2   VNHM1-1750          Rượu Champagne Autreau Brut Premier Cru  1750000.0   
3   VNHM1-1850    Rượu Champagne Autreau Reserve Brut Grand Cru  1850000.0   
4   VNTK1-2760             Rượu Champagne Ayala Blanc De Blancs  2760000.0   
5    VNTK-2294                 Rượu Champagne Ayala Brut Majeur  2294000.0   
6    VNTK-2734                 Rượu Champagne Ayala Brut Nature  2734000.0   
7    VNPE-1250             Rượu Champagne Bernard Lonclas Blanc  1250000.0   
8   VNWH1-4200  Rượu Champagne Billecart Salmon Blanc De Blancs  2709000.0   
9   VNWH1-2750     Rượu Champagne Billecart Salmon Brut Réserve  1895000.0   
10   VNWH-3900   Rượu Champagne Billecart Salmon Brut Sous Bois  3900000.0   
11    VNDL-552          Rượu Champagne Bisol BelStar Cuvee Rose   495000.0   
12  VNTK1-3867     Rượu Champagne Bollinger La Cote Aux Enfants  3867000.0   
13   VNTK-4920         Rượu Champagne Bollinger La Grande Année  4920000.0   
14   VNTK-5960    Rượu Champagne Bollinger La Grande Année Rose  5960000.0   
15   VNTK-3880                    Rượu Champagne Bollinger Rose  3880000.0   
16   VNTK-4134       Rượu Champagne Bollinger Rose Limited 2006  4134000.0   
17   VNTK-3294           Rượu Champagne Bollinger Special Cuvee  3294000.0   
18   VNPV-3958         Rượu Champagne Bottega Stardust Prosecco  3563000.0   
19   VNTL-1528               Rượu Champagne Canard Duchene Brut  1290000.0   

   Xuat_xu  Nong_do  Dung_tich                             Giong_nho  \
0     Pháp     12.5      750.0                            Chardonnay   
1     Pháp     12.0      750.0                            Chardonnay   
2     Pháp     12.0      750.0                            Chardonnay   
3     Pháp     12.0      750.0                 Chardonnay,Pinot Nero   
4     Pháp     12.0      750.0                            Chardonnay   
5     Pháp     12.0      750.0   Chardonnay,Pinot Noir,Pinot Meunier   
6     Pháp     12.0      750.0   Chardonnay,Pinot Noir,Pinot Meunier   
7     Pháp     12.0      750.0  Pinot Noir,Pinot Meunieur,Chardonnay   
8     Pháp     12.5      750.0                           Chardnonnay   
9     Pháp     12.5      750.0   Chardonnay,Pinot Noir,Pinot Meunier   
10    Pháp     12.5      750.0   Chardonnay,Pinot Noir,Pinot Meunier   
11       Ý     11.5      750.0                     Merlot,Pinot Nero   
12    Pháp     12.5      750.0                            Pinot Noir   
13    Pháp     12.0      750.0                 Pinot Noir,Chardonnay   
14    Pháp     14.0      750.0   Pinot Noir,Chardonnay,Pinot Meunier   
15    Pháp     12.0      750.0                 Chardonnay,Pinot Noir   
16    Pháp     12.0      750.0                 Chardonnay,Pinot Noir   
17    Pháp     12.0      750.0                 Chardonnay,Pinot Noir   
18       Ý     11.0      750.0                                 Glera   
19    Pháp     12.0      750.0   Pinot Meunier,Pinot Noir,Chardonnay   

          Mau_sac         Loai_ruou               Thuong_hieu Vung_lam_vang  
0           trắng    Rượu Champagne         Armand De Brignac           NaN  
1             NaN    Rượu Champagne           Arnaud Tabourin           NaN  
2             NaN  Rượu Vang Trắng                       NaN     Champagne  
3             NaN    Rượu Champagne                   Autreau           NaN  
4      vàng nhạt   Rượu Champagne                     Ayala           NaN  
5        vàng rơm   Rượu Champagne                     Ayala           NaN  
6        vàng rơm   Rượu Champagne                     Ayala           NaN  
7        vàng rơm    Rượu Champagne                       NaN           NaN  
8             NaN         Champagne       Champagne Billecart           NaN  
9        vàng rơm         Champagne       Champagne Billecart           NaN  

## Quay lại bước khám phá dữ liệu 

Với các cột "Xuat_xu","Giong_nho","Mau_sac","Thuong_hieu" ta cần tính:
- Tỉ lệ % (từ 0 đến 100) các giá trị thiếu 
- Số lượng các giá trị (các giá trị ở đây là các giá trị khác nhau và ta không xét giá trị thiếu): với cột mà ứng với câu hỏi dạng multichoice (ví dụ, cột "Giong_nho"), mỗi giá trị có thể chứa nhiều choice (các choice được phân tách bởi dấu phẩy)
- Tỉ lệ % (từ 0 đến 100) của mỗi giá trị được sort theo tỉ lệ % giảm dần (ta không xét giá trị thiếu, tỉ lệ là tỉ lệ so với số lượng các giá trị không thiếu)

Lưu kết quả vào DataFrame `cat_col_info_df`, trong đó: 
- Tên của các cột là tên của các cột không phải số trong `survey_df`
- Tên của các dòng là: "missing_ratio", "num_values", "value_ratios"  

Để dễ nhìn, tất cả các giá trị bạn đều làm tròn với 1 chữ số thập phân bằng phương thức `.round(1)`.

In [22]:
pd.set_option('display.max_colwidth', 100) # Để dễ nhìn
pd.set_option('display.max_columns', None) # Để dễ nhìn

temp = df[["Xuat_xu","Giong_nho","Mau_sac","Thuong_hieu"]]
items = ["Xuat_xu","Giong_nho","Mau_sac","Thuong_hieu"]


percentMissingList = []
for item in items:
   percentMissingList.append(round(df[item].apply(is_nan).tolist().count(1)/len(df) * 100,1))

columns = temp.columns.tolist()
row_labels = ["missing_ratio", "num_values", "value_ratios"]

num_values_list = []
for k in range (0,len(items)):
    s = temp[items[k]].tolist()
    s_split = []
    for i in range(0,len(s)):
        if is_nan(s[i]) == 0:
            s_split.append(re.split(', ',s[i]))

    l = pd.Series(s_split).explode().tolist()
    l = [*set(l)]
    num_values_list.append(len(l))
        

list_dic = []

for k in range (0,len(items)):
    s = temp[items[k]].tolist()
    s_split = []
    for i in range(0,len(s)):
        if is_nan(s[i]) == 0:
            s_split.append(re.split(', ',s[i]))

    l = pd.Series(s_split).explode()
    a = l.value_counts()
    dic = round(a/a.sum()*100,1).to_dict()
    list_dic.append(dic)       

data = [percentMissingList, num_values_list, list_dic]
cat_col_info_df = pd.DataFrame(data = data, index = row_labels, columns = columns)

In [23]:
cat_col_info_df

Xuat_xu  \
missing_ratio                                                                                                  0.5   
num_values                                                                                                      20   
value_ratios   {'Pháp': 37.4, 'Ý': 20.4, 'Chile': 16.1, 'ÚC': 7.3, 'Tây Ban Nha': 4.7, 'Mỹ': 4.2, 'Argentina': ...   

                                                                                                         Giong_nho  \
missing_ratio                                                                                                  7.7   
num_values                                                                                                     851   
value_ratios   {'Chardonnay': 9.9, 'Cabernet Sauvignon': 7.9, 'Pinot Noir': 5.7, 'Sauvignon Blanc': 5.3, 'Caber...   

                                                                                                           Mau_sac  \
missing_ratio                                                                                                 17.2   
num_values                                                                                                     248   
value_ratios   {'đỏ ruby': 38.2, 'vàng rơm': 15.4, 'đỏ': 9.3, 'đỏ đậm': 7.7, 'vàng nhạt': 5.9, 'đỏ ruby đậm': 1...   

                                                                                                       Thuong_hieu  
missing_ratio                                                                                                 29.3  
num_values                                                                                                    1415  
value_ratios   {'Concha Y Toro': 1.3, 'Borie Manoux': 0.7, 'Miguel Torres': 0.7, 'Santa Rita': 0.6, 'Frescobald...

### Xử lý dữ liệu bị thiếu

Tổng kết lại một chút về tỉ lệ thiếu dữ liệu của từng cột

In [24]:
df.agg(lambda x : x.isna().mean()*100).sort_values(ascending = False)


Thuong_hieu      29.317656
Mau_sac          17.232074
Vung_lam_vang    11.603701
Giong_nho         7.729375
Loai_ruou         6.861989
Ma_sp             1.272167
Nong_do           0.578258
Ten_sp            0.520432
Gia               0.520432
Xuat_xu           0.520432
Dung_tich         0.520432
dtype: float64

Ta có thể thấy là 3 cột có tỉ lệ thiếu cao nhất có tỉ lệ lớn hơn 10% so tổng số lượng dữ liệu, cá biệt là cột Thuong_hieu có tỉ lệ thiếu lên đến 28%. Nếu chọn giải pháp xoá đi các dòng bị thiếu dữ liệu thì sẽ ảnh hưởng không nhỏ đến tính khách quan và chuẩn xác của tập dữ liệu. Cuối cùng, nhóm đã chọn phương án là lấp các ô bị thiếu ở cột Thuong_hieu bằng cách dựa vào cột Vung_lam_vang. Bởi vì một thương hiệu rượu thường sẽ đi liền với một vùng làm rượu riêng của họ. Ta sẽ xác định xem với từng vùng làm rượu thì thương hiệu nào xuất hiện nhiều nhất, từ đó lấy thương hiệu này làm đại diện. 

In [25]:
def is_empty(s):
  '''
    Return a boolean series based on whether the cell is empty or not.
    A cell is empty if it has either NA value or empty string.
  '''
  return ((s.isna()) | (s == ''))
# Get rows that neither cells in Thuong_hieu nor Vung_lam_vang columns are empty
filter_df = df.loc[~(is_empty(df['Thuong_hieu']) | is_empty(df['Vung_lam_vang']))].reset_index(drop = True)
place_brand_df = filter_df.groupby(['Vung_lam_vang', 'Thuong_hieu'])['Thuong_hieu'].count().to_frame(name = 'count')
# Get indices of brand having most count by each place of origin
idx = place_brand_df.groupby(level = 0)['count'].idxmax()
place_brand_df = place_brand_df.loc[idx]
place_brand_df = place_brand_df.sort_values(by = 'count', ascending = False)
place_brand_df = place_brand_df.reset_index()[['Vung_lam_vang','Thuong_hieu']]
place_brand_df

Vung_lam_vang        Thuong_hieu
0                 Bordeaux       Borie Manoux
1                  Mendoza           Trivento
2         Aconcagua Valley          Errazuriz
3                Bourgogne        Louis Jadot
4           Central Valley      Concha Y Toro
..                     ...                ...
366        Langhorne Creek           McGuigan
367   Languedoc Roussillon      Bories Azeaux
368               Lenswood           Henschke
369   Limori Valley, Chile            Mancura
370  Đảo Sardinia – Italia  Conte di Campiano

[371 rows x 2 columns]

In [26]:
def fill_empty_brand(row, place_brand_df):
    '''
    Fill empty cells on Thuong_hieu and Vung_lam_vang columns based on the place_brand_df
    as long as cells in both columns are not empty
    '''
    brand = row['Thuong_hieu']
    place_origin = row['Vung_lam_vang']
    #Ignore if both values are na value
    if pd.isnull(brand) and pd.isnull(place_origin):
        return row
    if brand == '' or pd.isnull(brand):
        # Get the corresponding brands
        replacing_brand = place_brand_df.loc[place_brand_df['Vung_lam_vang'] == place_origin, 'Thuong_hieu']
        if len(replacing_brand) > 0:
            # Only take the first item
            row['Thuong_hieu'] = replacing_brand.loc[replacing_brand.index[0]]
    elif place_origin == '' or pd.isnull(place_origin):
        # Get the corresponding places
        replacing_place = place_brand_df.loc[place_brand_df['Thuong_hieu'] == brand, 'Vung_lam_vang']
        if len(replacing_place) > 0:
            # Only take the first item
            row['Vung_lam_vang'] = replacing_place.loc[replacing_place.index[0]]
    return row

clean_df = df.apply(lambda x : fill_empty_brand(x,place_brand_df), axis = 1)
clean_df

Ma_sp                                         Ten_sp        Gia  \
0       WVN-9400         Rượu Champagne Armand De Brignac Blanc  9400000.0   
1      VNHM-1350                 Rượu Champagne Arnaud Tabourin  1350000.0   
2     VNHM1-1750        Rượu Champagne Autreau Brut Premier Cru  1750000.0   
3     VNHM1-1850  Rượu Champagne Autreau Reserve Brut Grand Cru  1850000.0   
4     VNTK1-2760           Rượu Champagne Ayala Blanc De Blancs  2760000.0   
...          ...                                            ...        ...   
5183   VNMH2-310    Rượu Vang Santa Carolina Cabernet Sauvignon   300000.0   
5184    VNPV-366           Rượu Vang Sunrise Cabernet Sauvignon   330000.0   
5185   VNWH4-328          Rượu Vang Trắng Deakin Estate Moscato   318000.0   
5186  VNWH1- 674            Rượu Vang ÚC Yalumba Organic Shiraz   660000.0   
5187  VNWH8- 595         Rượu Vang Yalumba Christobel’s Moscato   575000.0   

     Xuat_xu  Nong_do  Dung_tich              Giong_nho     Mau_sac  \
0       Pháp     12.5      750.0             Chardonnay       trắng   
1       Pháp     12.0      750.0             Chardonnay         NaN   
2       Pháp     12.0      750.0             Chardonnay         NaN   
3       Pháp     12.0      750.0  Chardonnay,Pinot Nero         NaN   
4       Pháp     12.0      750.0             Chardonnay  vàng nhạt   
...      ...      ...        ...                    ...         ...   
5183   Chile     13.5      750.0                    NaN         NaN   
5184   Chile     13.5      750.0     Cabernet Sauvignon     đỏ ruby   
5185      ÚC      6.5      750.0                Moscato         NaN   
5186      ÚC     13.5      750.0                 Shiraz  đỏ ánh tím   
5187      ÚC     12.5      750.0                Moscato    vàng rơm   

             Loai_ruou        Thuong_hieu    Vung_lam_vang  
0       Rượu Champagne  Armand De Brignac              NaN  
1       Rượu Champagne    Arnaud Tabourin              NaN  
2     Rượu Vang Trắng          Delamotte        Champagne  
3       Rượu Champagne            Autreau              NaN  
4      Rượu Champagne              Ayala              NaN  
...                ...                ...              ...  
5183               NaN                NaN              NaN  
5184      Rượu Vang Đỏ      Concha Y Toro   Central Valley  
5185   Rượu Vang Trắng      Deakin Estate         Victoria  
5186      Rượu vang đỏ            Yalumba  South Australia  
5187    Rượu Vang Ngọt            Yalumba   Barossa Valley  

[5188 rows x 11 columns]

In [27]:
clean_df.agg(lambda x : x.isna().mean()*100)

Ma_sp             1.272167
Ten_sp            0.520432
Gia               0.520432
Xuat_xu           0.520432
Nong_do           0.578258
Dung_tich         0.520432
Giong_nho         7.729375
Mau_sac          17.232074
Loai_ruou         6.861989
Thuong_hieu       9.888204
Vung_lam_vang    10.890517
dtype: float64

Sau khi điền vào các ô bị thiếu, tỉ lệ thiếu ở 2 cột Thuong_hieu và Vang_lam_vang giảm đi đáng kể chỉ còn 9-10%. Lúc này ta có thể tiến hành loại bỏ đi các dòng có ô còn bỏ trống.

In [28]:
#Drop rows having NA value
clean_df = clean_df.dropna()
cate_cols = clean_df.select_dtypes(include = ['object']).columns.tolist()
empty_string_count = clean_df[cate_cols].agg(lambda x : (x == '').sum())
empty_string_count

Ma_sp             0
Ten_sp            0
Xuat_xu           0
Giong_nho        13
Mau_sac           0
Loai_ruou         0
Thuong_hieu       0
Vung_lam_vang    49
dtype: int64

Có 4 cột xuất hiện string rỗng. Ta sẽ xoá những dòng có string rỗng dựa trên 4 cột này

In [29]:
empty_string_cols = empty_string_count[empty_string_count > 0].index
empty_string_rows = clean_df.apply(lambda x : any(x[col] == '' for col in empty_string_cols), axis = 1)
clean_df = clean_df[~empty_string_rows].reset_index(drop = True)
clean_df.head(10)

Ma_sp                                         Ten_sp        Gia  \
0   VNPV-3958       Rượu Champagne Bottega Stardust Prosecco  3563000.0   
1   VNPV-9463                       Rượu Champagne Carbon F1  8517000.0   
2  VNPV1-1823  Rượu Champagne Charles Mignon Premium Reserve  1641000.0   
3    VNPV-836              Rượu Champagne De Castellane Brut   753000.0   
4   VNPV-4233       Rượu Champagne Delamotte Blanc De Blancs  3810000.0   
5   VNPV-1956                  Rượu Champagne Delamotte Brut  1761000.0   
6   VNPV-3667                  Rượu Champagne Delamotte Rose  2956000.0   
7   VNPV-1976     Rượu Champagne Duval Leroy Thierry Drapeau  1779000.0   
8   VNTK3-480            Rượu Champagne JP Chenet Brut Blanc   480000.0   
9   VNBMF-744     Rượu Champagne JP Chenet Divine Chardonnay   750000.0   

  Xuat_xu  Nong_do  Dung_tich                            Giong_nho  \
0       Ý     11.0      750.0                                Glera   
1    Pháp     12.0      750.0  Chardonnay,Pinot meunier,Pinot Noir   
2    Pháp     13.5      750.0  Chardonnay,Pinot meunier,Pinot Noir   
3    Pháp     11.0      375.0                           Chardonnay   
4    Pháp     12.5      750.0                           Chardonnay   
5    Pháp     12.0      750.0                           Chardonnay   
6    Pháp     12.0      750.0  Chardonnay,Pinot Noir,Pinot Meunier   
7    Pháp     12.5      750.0                           Chardonnay   
8    Pháp     13.0      750.0                           Chardonnay   
9    Pháp     12.5      750.0                           Chardonnay   

         Mau_sac       Loai_ruou            Thuong_hieu Vung_lam_vang  
0  vàng rơm nhạt  Rượu Champagne                Bottega        Veneto  
1          trắng  Rượu Champagne              Delamotte     Champagne  
2           vàng  Rượu Champagne         Charles Mignon     Champagne  
3      xanh nhạt  Rượu Champagne          De Castellane     Champagne  
4    trắng trong  Rượu Champagne              Delamotte     Champagne  
5           vàng  Rượu Champagne              Delamotte     Champagne  
6           hồng  Rượu Champagne              Delamotte     Champagne  
7           vàng  Rượu Champagne  Champagne Duval Leroy     Champagne  
8        đỏ đậm  Rượu Vang Nổ              JP Chenet     Pays d’Oc  
9       vàng rơm  Rượu Vang Nổ              JP Chenet     Pays d’Oc

In [30]:
row_num = clean_df.shape[0]
col_num =  clean_df.shape[1]
print('Số dòng: {}\nSố cột: {}'.format(row_num,col_num))

Số dòng: 3934
Số cột: 11


Số dòng lúc này chỉ còn lại là 3950 dòng so với 5598 dòng ban đầu, dữ liệu bị hao hụt không quá đáng kể.

In [31]:
clean_df[clean_df['Vung_lam_vang'].apply(lambda x : len(x) >= 50)]

Ma_sp                                  Ten_sp       Gia Xuat_xu  \
1726  VNYV-320  Rượu Vang Hera Mitologia Nero Di Troia  320000.0       Ý   

      Nong_do  Dung_tich      Giong_nho  Mau_sac     Loai_ruou Thuong_hieu  \
1726     13.5      750.0  Nero Di Troia  đỏ ruby  Rượu Vang Đỏ   Mitologia   

                                                                                            Vung_lam_vang  
1726  chính là cái “nôi” của những dòng rượu vang nổi tiếng của đất nước Ý hình chiếc ủng xinh xă...

Có một dòng ở cột vùng làm rượu chứa thông tin rất dài, có thể là do gặp lỗi nhỏ trong quá trình thu nhập dữ liệu. Ta sẽ xoá đi dòng này

In [32]:
clean_df = clean_df.drop(index = 1726)

Lưu dữ liệu đã qua tiền xử lý vào file dữ liệu mới.

In [33]:
clean_df.to_csv(os.path.join(os.getcwd(), 'dataset', 'champagne_info_cleaned.csv'))

### Trực quan một vài biểu đồ để có một cái nhìn tổng quan hơn về dữ liệu

Lấy ra các cột kiểu numeric

In [34]:
numeric_cols = df.select_dtypes('number').columns.tolist()
numeric_cols

['Gia', 'Nong_do', 'Dung_tich']

Biểu đồ boxplot của từng cột có kiểu dữ liệu numeric

In [ ]:
pretty_labels = {
  'Gia' : 'Giá (Triệu)',
  'Nong_do' : 'Nồng độ (%)',
  'Dung_tich' : 'Dung tích (ml)'
}
fig, axes = plt.subplots(3,1, figsize = (16,9))
for i,ax in enumerate(axes.flat):
  sns.boxplot(data = df, x = numeric_cols[i], ax = ax)
  ax.set_xlabel(pretty_labels[numeric_cols[i]])
  if i == 0:
    ax.set_xticklabels(['{}'.format(x) for x in ax.get_xticks()/1e6])

plt.tight_layout()

Nhìn chung, giá trị của các cột dữ liệu kiểu số đều không phân bố rộng mà chỉ tập trung vào một khoảng giá trị nhỏ cũng như xuất hiện nhiều ngoại lệ.

Xem xét thử mối quan hệ giữa giá với nồng độ và dung tích

In [ ]:
fig, axes = plt.subplots(1,2, figsize = (12,6))
sns.scatterplot(data = df, x = 'Gia', y = 'Nong_do', ax = axes[0])
sns.scatterplot(data = df, x = 'Gia', y = 'Dung_tich', ax = axes[1])
axes[0].set_xlabel(pretty_labels['Gia'])
axes[0].set_xticklabels(['{}'.format(x) for x in axes[0].get_xticks()/1e6])
axes[0].set_ylabel(pretty_labels['Nong_do'])
axes[1].set_xlabel(pretty_labels['Gia'])
axes[1].set_xticklabels(['{}'.format(x) for x in axes[0].get_xticks()/1e6])
axes[1].set_ylabel(pretty_labels['Dung_tich'])

Có vẻ là giá rượu không hề có mối liên hệ gì với nồng độ hay dung tích

Những quốc gia được nhập rượu về nhiều nhất

In [ ]:
origin_count_df = df['Xuat_xu'].value_counts().to_frame(name = 'Số lượng sản phẩm').reset_index()
origin_count_df = origin_count_df.rename(columns = {'index' : 'Xuất xứ'})
fig, ax = plt.subplots(figsize = (12,8))
sns.barplot(data = origin_count_df, x = 'Số lượng sản phẩm', y = 'Xuất xứ', orient = 'h', palette = 'muted', ax = ax)